core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -176,7 +176,8 @@ class KafkaApis(val requestChannel: RequestChannel,
       case e: FatalExitError => throw e
       case e: Throwable => handleError(request, e)
     } finally {
-      request.apiLocalCompleteTimeNanos = time.nanoseconds
+      if (request.apiLocalCompleteTimeNanos < 0)
+        request.apiLocalCompleteTimeNanos = time.nanoseconds
```

Comment:
This actually seems like an intuitive time to set the local complete time. I guess the reason we set it in `RequestChannel.Response` also is that we are trying to set it consistently with `responseCompleteTimeNanos`. I can't really think of a better solution than what you have here, but it might at least be helpful to add a comment of explanation?

Reply:
I was looking at this code and the way it works today is super confusing. :) I think this change makes sense. However, I was curious how you noticed this. Was there a delay in getting to this point somehow? In theory, the value at this point should be very similar to the value we set during processing.

Reply:
From what I remember, the value was close enough that I didn't consider it worth looking into. I encountered it when I added debug metrics to further subdivide the local time, and there appeared to be lost time, i.e. the sum wasn't adding to the whole.

Reply:
Makes sense, thanks.